esp/esdllib/esdl_script.cpp

```diff
@@ -136,26 +141,133 @@ class CEsdlTransformOperationParameter : public CEsdlTransformOperationVariable
+{
+protected:
+    StringAttr m_name;
+    Owned<ICompiledXpath> m_xpath_name;
+    Owned<ICompiledXpath> m_select;
+
+public:
+    CEsdlTransformOperationSetSectionAttributeBase(IPropertyTree *tree, const StringBuffer &prefix, const char *attrName) : CEsdlTransformOperationBase(tree, prefix)
+    {
+        if (m_traceName.isEmpty())
+            m_traceName.set(tree->queryProp("@name"));
+        if (!isEmptyString(attrName))
+            m_name.set(attrName);
+        else
+        {
+            m_name.set(tree->queryProp("@name"));
+
+            const char *xpath_name = tree->queryProp("@xpath_name");
+            if (!isEmptyString(xpath_name))
+                m_xpath_name.setown(compileXpath(xpath_name));
+            else if (m_name.isEmpty())
+                esdlOperationError(ESDL_SCRIPT_MissingOperationAttr, m_tagname, "without name", m_traceName, !m_ignoreCodingErrors); //don't mention value, it's deprecated
+        }
+
+        const char *select = tree->queryProp("@select");
+        if (isEmptyString(select))
+            esdlOperationError(ESDL_SCRIPT_MissingOperationAttr, m_tagname, "without select", m_traceName, !m_ignoreCodingErrors); //don't mention value, it's deprecated
+        m_select.setown(compileXpath(select));
+    }
+
+    virtual void toDBGLog() override
+    {
+#if defined(_DEBUG)
+        DBGLOG(">%s> %s(name(%s), select('%s'))", m_traceName.str(), m_tagname.str(), (m_xpath_name) ? m_xpath_name->getXpath() : m_name.str(), m_select->getXpath());
+#endif
+    }
+
+    virtual ~CEsdlTransformOperationSetSectionAttributeBase(){}
+
+    virtual const char *getSectionName() = 0;
+
+    virtual bool process(IEsdlScriptContext * context, IPropertyTree *content, IXpathContext * xpathContext) override
+    {
+        if ((!m_name && !m_xpath_name) || !m_select)
+            return false; //only here if "optional" backward compatible support for now (optional syntax errors aren't actually helpful)
+        try
+        {
+            StringBuffer name;
+            if (m_xpath_name)
+                xpathContext->evaluateAsString(m_xpath_name, name);
+            else
+                name.set(m_name);
+
+            StringBuffer value;
+            xpathContext->evaluateAsString(m_select, value);
+            context->setAttribute(getSectionName(), name, value);
+        }
+        catch (IException* e)
+        {
+            int code = e->errorCode();
+            StringBuffer msg;
+            e->errorMessage(msg);
+            e->Release();
+            esdlOperationError(code, m_tagname, msg, m_traceName, !m_ignoreCodingErrors);
+        }
+        catch (...)
+        {
+            esdlOperationError(ESDL_SCRIPT_Error, m_tagname, "unknown exception processing", m_traceName, !m_ignoreCodingErrors);
+        }
+        return false;
+    }
+};
+
+class CEsdlTransformOperationStoreValue : public CEsdlTransformOperationSetSectionAttributeBase
```

Comment:
Would there be value allowing more complex values to be stored? If we think we might want to store more structured data in the future, we might not want to rely on attributes now. I'm not saying we will. I'm just asking if it was considered.

Reply:
yes I did think about it.  I'm just leaving the actual content of the section for future consideration.  If we need something more than key value pairs I'd like to understand it before adding support... but will do so if we need it.

Also it will be more valuable when we have better ways of building structures.  Perhaps after we've moved away from IPtree on the right hand side.

Reply:
I guess I was thinking that more complex data would suggest child elements. We could use child elements with name and value for the current key value pairs, then for more complex data replace the value attribute with content. Just thought it might be more consistent when we grow into it. Either way works.